File: tool/coding/frontend/pixel_display.js
// ========== pixel_display.js - 工具定义 ==========
// 朱峰社区智能体无限 5.0.0 - 独立工具文件
// 自动生成自 4.1.8 tools.js（split_tools.js V3）
// 执行逻辑: tools/executor.js 中对应 case 分支

window.Tools = window.Tools || { allTools: {} };
if (!window.Tools.allTools['pixel_display']) {
    window.Tools.allTools['pixel_display'] = {
    pixel_display: {
        type: 'function',
        function: {
            name: 'pixel_display',
            description: '像素显示器。向左下角面板发送PXL像素图/动画。静态:WxHB:RLE 动画:WxHB F帧数:R1|R2 动画带帧率:WxHB F帧数@fps:R1|R2。RLE从0(黑)开始交替计数。fps默认4,动画循环播放。',
            parameters: {
                type: 'object',
                properties: {
                    action: { type: 'string', enum: ['show', 'clear', 'status'], description: 'show=显示,clear=清除,status=查状态' },
                    title: { type: 'string', description: '标题(可选)' },
                    data: { type: 'string', description: 'PXL数据。静态:16x16B:36,2,3,2,8,4,1,4,... 动画:16x16B F2:36,2,...|16,16,...  带fps:16x16B F4@8:R1|R2|R3|R4' },
                    fps: { type: 'integer', description: '帧率(可选,默认4。也可在data中用@fps指定)' }
                },
                required: ['action']
            }
        }
    },

    // ===== 正则搜索工具（始终正则模式，支持捕获组提取） =====
    };
}

File: tools/coding/backend/pixel_display.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""pixel_display - 像素显示器"""

import os
import json
import time
from tools.coding.backend.base import ToolContext

TOOL_NAME = 'pixel_display'

CATEGORY = 'pixel_display'
KEY_LATEST = 'latest'


def handle(body, ctx):
    """处理工具请求"""
    try:
        action = body.get('action', 'status')

        if action == 'show':
            _do_show(body, ctx)
        elif action == 'clear':
            _do_clear(ctx)
        elif action == 'status':
            _do_status(ctx)
        else:
            ctx.send_json({'ok': False, 'error': f'未知操作: {action}'})
    except Exception as e:
        ctx.send_json({'ok': False, 'error': str(e)})


def _do_show(body, ctx):
    """存储PXL数据到app_data，前端轮询读取显示"""
    title = body.get('title', '')
    data = body.get('data', '')
    fps = body.get('fps', 2)
    now = int(time.time() * 1000)

    if not data:
        ctx.send_json({'ok': False, 'error': '需要提供 data 参数'})
        return

    display_data = {
        'title': title,
        'data': data,
        'fps': fps,
        'updated_at': now
    }

    with ctx.db_lock:
        conn = ctx.get_db()
        try:
            # 检查是否已有记录
            row = conn.execute(
                'SELECT id FROM app_data WHERE category=? AND key=?',
                [CATEGORY, KEY_LATEST]
            ).fetchone()

            if row:
                conn.execute(
                    'UPDATE app_data SET value=?, updated_at=? WHERE category=? AND key=?',
                    [json.dumps(display_data, ensure_ascii=False), now, CATEGORY, KEY_LATEST]
                )
            else:
                conn.execute(
                    'INSERT INTO app_data (category, key, value, created_at, updated_at) VALUES (?, ?, ?, ?, ?)',
                    [CATEGORY, KEY_LATEST, json.dumps(display_data, ensure_ascii=False), now, now]
                )
            conn.commit()
        finally:
            conn.close()

    ctx.send_json({'ok': True, 'message': '像素显示数据已更新', 'display': display_data})


def _do_clear(ctx):
    """删除app_data中的pixel_display记录"""
    with ctx.db_lock:
        conn = ctx.get_db()
        try:
            conn.execute(
                'DELETE FROM app_data WHERE category=? AND key=?',
                [CATEGORY, KEY_LATEST]
            )
            conn.commit()
        finally:
            conn.close()

    ctx.send_json({'ok': True, 'message': '像素显示器已清空'})


def _do_status(ctx):
    """返回当前状态"""
    with ctx.db_lock:
        conn = ctx.get_db()
        try:
            row = conn.execute(
                'SELECT value, updated_at FROM app_data WHERE category=? AND key=?',
                [CATEGORY, KEY_LATEST]
            ).fetchone()
        finally:
            conn.close()

    if row:
        try:
            data = json.loads(row['value'])
        except Exception:
            data = {'raw': row['value']}
        ctx.send_json({'ok': True, 'active': True, 'display': data, 'updated_at': row['updated_at']})
    else:
        ctx.send_json({'ok': True, 'active': False, 'display': None})
